DAOS-623 actions: Improve github actions support.

ashleypittman opened the pull request:
Add another build test to github-actions.  This doesn't
modify the existing code, but intends to replace it once
stable.

Use the github cache to re-use docker images for speed.
Do a single image create, then re-use this for different
compilers etc.
Run NLT in docker, and report server leaks as annotations.
Remove PSM2 from the Docker builds.

This is enabled to build centos.8 with gcc/clang, and to
build and test with ubuntu/clang but can be readily
extended to the rest of the matrix once landed and stable
without adding additional time or complexity.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented on utils/run_in_ga.sh:
```diff
@@ -0,0 +1,36 @@
+echo ::endgroup::
+
+echo ::group::Build type debug.
+$SCONS --jobs 10 PREFIX=/opt/daos COMPILER=$COMPILER TARGET_TYPE=release BUILD_TYPE=debug
```
(style)  line over 80 characters
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/run_in_ga.sh:
```diff
@@ -0,0 +1,36 @@
+
+echo ::group::Container copy test
+# DAOS-7440
+export LD_LIBRARY_PATH=/opt/daos/lib:/opt/daos/lib64:/opt/daos/bin/../lib64/daos_srv/../../prereq/release/spdk/lib/:$LD_LIBRARY_PATH
```
(style)  line over 80 characters

daosbuild1 commented on utils/run_in_ga_wrapper.sh:
```diff
@@ -0,0 +1,5 @@
+
+IMAGE=$(docker images -all --filter label=DAOS=true --quiet | tail -n 1)
+
+docker run --name build-post --mount type=tmpfs,destination=/mnt/daos,tmpfs-mode=1777 --env COMPILER $IMAGE ./utils/run_in_ga.sh
```
(style)  line over 80 characters
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/node_local_test.py:
```diff
@@ -278,6 +285,10 @@ def add(self, line, sev, message, cat=None, mtype=None):
         self._flush()
+        if self.post:
+            # https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions
+          print('::warning file={},line={},::{}, {}'.format(line.filename, line.lineno, self.check,
```
(pylint-line-too-long) Line too long (99/80)
(pylint-bad-indentation) Bad indentation. Found 10 spaces, expected 12

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5694/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[utils/node_local_test.py:2738](https://github.com/daos-stack/daos/blob/5423299a95cca980b4afa2a2f0ce5b2307163f1b/utils/node_local_test.py#L2738):
(pylint-too-many-branches) Too many branches (26/25)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5694/1/execution/node/98/log

daosbuild1 commented on utils/run_in_ga.sh:
```diff
@@ -0,0 +1,36 @@
+echo ::endgroup::
+
+echo ::group::Build type debug.
+$SCONS --jobs 10 PREFIX=/opt/daos COMPILER="$COMPILER" TARGET_TYPE=release BUILD_TYPE=debug
```
(style)  line over 80 characters

daosbuild1 commented on utils/run_in_ga_wrapper.sh:
```diff
@@ -0,0 +1,5 @@
+
+IMAGE=$(docker images --all --filter label=DAOS=true --quiet | tail -n 1)
+
+docker run --name build-post --mount type=tmpfs,destination=/mnt/daos,tmpfs-mode=1777 --env COMPILER "$IMAGE" ./utils/run_in_ga.sh
```
(style)  line over 80 characters

daosbuild1 commented on utils/node_local_test.py:
```diff
@@ -278,6 +285,12 @@ def add(self, line, sev, message, cat=None, mtype=None):
         self._flush()
+        if self.post:
+          # https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions
+          print('::warning file={},line={},::{}, {}'.format(line.filename,
```
(pylint-bad-indentation) Bad indentation. Found 10 spaces, expected 12

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5694/2/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[utils/node_local_test.py:2740](https://github.com/daos-stack/daos/blob/d7c3e8182ebdea2a08e95804ea5c5225b066e1b6/utils/node_local_test.py#L2740):
(pylint-too-many-branches) Too many branches (26/25)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5694/2/execution/node/99/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/node_local_test.py:2740](https://github.com/daos-stack/daos/blob/108fcba6ded4f19a4cb093c7fb77bede390b3b7b/utils/node_local_test.py#L2740):
(pylint-too-many-branches) Too many branches (26/25)

daosbuild1 commented:
Test stage Unit Test with memcheck completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5694/3/display/redirect

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5694/3/display/redirect

daosbuild1 commented:
Test stage NLT completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5694/3/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/node_local_test.py:2740](https://github.com/daos-stack/daos/blob/4cbe341cb07eb039fe2110070d876a8a892df2ec/utils/node_local_test.py#L2740):
(pylint-too-many-branches) Too many branches (26/25)

jolivier23 commented:
Do we not still need to keep opa?  Have we replaced all of the legacy hardware?

jolivier23 commented on utils/sl/components/__init__.py:
```diff
@@ -66,13 +66,6 @@ def check(self, name):
         self.not_installed.append(name)
         return False
 
-def exclude(reqs, name, use_value, exclude_value):
```
If we remove this, do we also need to remove the EXCLUDE Variable?   Presently, it is only used for psm2 exclusion

ashleypittman replied:
Possibly, you mean as a command line option?  I can't rule out needing this function again at some point so I was on the fence about leaving this in.

jolivier23 replied:
yes, this implements the command line option by checking if psm2 is included on the command line.

ashleypittman replied:
This bit isn't actually required for the rest of the functionality, I only included it as Johann had said that he thought we'd already remove it.

I did try testing with a different distro at one point, and psm2 doesn't compile there which is why it got removed, but that's no longer part of this PR.

jolivier23 reviewed (changes requested):
Adding request changes to get it off my inbox until question(s) are addressed.

ashleypittman commented:
> Do we not still need to keep opa? Have we replaced all of the legacy hardware?

I didn't change anything with openpa did I?  I've no idea if we still need it or not, but it's outside the scope of this change.

jolivier23 commented:
> > Do we not still need to keep opa? Have we replaced all of the legacy hardware?
> 
> I didn't change anything with openpa did I? I've no idea if we still need it or not, but it's outside the scope of this change.

I mean opa the fabric, not openpa the atomic library.  Actually opa2 which is what psm2 is for.   I think a lot of our internal nodes may still have these cards but I'm not sure.  @johannlombardi ?

ashleypittman commented:
> > > Do we not still need to keep opa? Have we replaced all of the legacy hardware?
> > 
> > 
> > I didn't change anything with openpa did I? I've no idea if we still need it or not, but it's outside the scope of this change.
> 
> I mean opa the fabric, not openpa the atomic library. Actually opa2 which is what psm2 is for. I think a lot of our internal nodes may still have these cards but I'm not sure. @johannlombardi ?

I've removed that part of the PR, it was incidental to the rest of the work.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/node_local_test.py:2740](https://github.com/daos-stack/daos/blob/6eee8b7512fa1908ca7b07dd662dc54b01ea1933/utils/node_local_test.py#L2740):
(pylint-too-many-branches) Too many branches (26/25)
